doc/docs/Python_Tutorials/Adjoint_Solver.md

```diff
@@ -2,13 +2,704 @@
+adjoint solver including worst-case (minimax) optimization, multiple
+objective functions, and constraints on the minimum line width and
+line spacing. The design problem involves a broadband waveguide mode
+converter with minimum feature size in 2D. This is based on
+[M.F. Schubert et al., ACS Photonics, Vol. 9, pp. 2327-36,
+(2022)](https://doi.org/10.1021/acsphotonics.2c00313).
+
+The mode converter must satisfy two separate design objectives: (1)
+minimize the reflectance $R$ into the fundamental mode of the input
+port ($\mathcal{S}_{11}$) and (2) maximize the transmittance $T$ into
+the second-order mode of the output port ($\mathcal{S}_{21}$). There
+are different ways to define this multi-objective and multi-wavelength
+optimization. The approach taken here is *worst-case* optimization
+whereby we minimize the maximum (the worst case) of $R$ and $1-T$
+across all six wavelengths (in the O-band for
+telecommunications). This is known as *minimax* optimization.
+
+The challenge with minimax optimization is that the $\max$ objective
+function is not everywhere differentiable. This property would seem to
+preclude the use of gradient-based optimization algorithms for this
+problem which involves 12 independent functions ($R$ and $1-T$ for
+each of six wavelengths). Fortunately, there is a workaround: the
+problem can be reformulated as a differentiable problem by introducing
+a dummy varaiable $t$ (the
+[epigraph](https://en.wikipedia.org/wiki/Epigraph_(mathematics))) and
+adding each independent function as a new nonlinear constraint. See
+the [NLopt
+documentation](https://nlopt.readthedocs.io/en/latest/NLopt_Introduction/#equivalent-formulations-of-optimization-problems)
+for an overview of this approach.
+
+In this example, we use a minimum feature size of 150 nm for the line
+width *and* line spacing. The implementation of these constraints is
+based on [A.M. Hammond et al., Optics Express, Vol. 29, pp. 23916-38
+(2021)](https://doi.org/10.1364/OE.431188).
+
+There are five important items to note in the set up of the
+optimization problem:
+
+- The lengthscale constraint is activated only in the final epoch. The
+  initial design of the final epoch must be binary.
+
+- The initial value of the epigraph variable of the final epoch (in
+  which the linewidth constraint is imposed) should take into account
+  the value of the constraint itself to ensure a feasible starting
+  point.
+
+- The edge of the design region is padded by a filter radius (rather
+  than e.g., a single pixel) to produce measured lengthscales of the
+  final design that are consistent with the imposed constraint.
+
+- The hyperparameters of the lengthscale constraint function (`glc` in
+  the script below), need to be chosen carefully to produce final
+  designs which do not significantly degrade the performance of the
+  unconstrained designs at the start of the final epoch.
+
+- Damping of the design weights is used for the early epochs in which
```

Comment:
I'm not sure I would say the weights are being dampened (the weights are non complex)...

Rather, we introduce a new damping term via a material conductivity which *depends* on the weights meant to encourage binarization.